Editor: Add sessionStorage autosave mechanism

aduth opened the pull request:
Closes #7367
Related: #6322

_Some edits by @mcsf_

This pull request seeks to implement periodic ~localStorage~ sessionStorage autosave backups for posts. It reuses and extends the behavior of the existing server-persisted autosave component, leveraging sessionStorage when available to save data on a more frequent basis.

![backup](https://user-images.githubusercontent.com/1779930/61158816-1688f680-a4c8-11e9-9d64-088206e276db.gif)

**Implementation notes:**

~This was designed to be compatible with localStorage backup mechanism implemented in the classic editor (WordPress 4.9.x and older), hence the selection of the localStorage key `'post_' + postId`.~

This influenced the choice to not use the existing [data persistence behavior](https://github.com/WordPress/gutenberg/tree/master/packages/data/src/plugins/persistence), along with a desire to avoid creating saves until the interval has elapsed (vs. on every state change).

Even with the delay, we may want to explore using [`requestIdleCallback`](https://developers.google.com/web/updates/2015/08/using-requestidlecallback) in addition to the timeout interval, since serialization can be a very non-performant operation, depending on the size of the post.

**Testing Instructions:**

The autosaves occur every 15 seconds after the last change has occurred. Verify that upon waiting this duration and reloading the page (dismissing the prompt), you see a restore option to restore the title, content, and excerpt of the post you had been working on.

Verify that the prompt is not shown if you proceed to make edits to the post and save (regardless of whether you choose to restore or dismiss the notice).

**Checklist:** (Added 2019-08-30)

- [x] Always run the autosave procedure every 15 seconds, not 15 seconds after the last change
- ~Still try to save to storage only if a change has actually occurred~
- [x] Explore whether requestIdleCallback is a reasonable improvement to avoid blocking operation
- [x] Add some more through inline code documention (e.g. lastDidSave)
- [x] Convert to use sessionStorage as a way to avoid accumulating storage and potential for lingering storage from other users
- ~Should confirm that this data is in-fact purged upon logging out, or that an equivalent measure is taken to avoid access by other users even within the same storage~
- [x] Fix `useAutosavePurge`, per https://github.com/WordPress/gutenberg/pull/16490#discussion_r324427512

aduth committed:
Editor: Add localStorage autosave mechanism

swissspidy commented:
Sounds like a good first step. I'd love to see using service workers for this though, as it makes for a more robust experience when offline.

aduth commented:
@swissspidy Do you see service workers as an alternative to this approach, or a supplement? It's been some time since I've worked with service workers, so any resources you can share would be appreciated.

swissspidy commented:
Using serve workers would be a supplement, especially because they require HTTPS. In #6322 @westonruter actually shared some very good ideas on how they could enhance the editing experience when offline.

For example, it can cache assets and even REST API responses locally. So when you are offline, you could still refresh the page, select tags, and write content as usual. You could even go and edit some totally different post and have all the (cached) data at your fingertips. And thanks to background sync, any attempt to save the post will be recorded and run as soon as you get online again.

While https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API covers the fundamental basics of service workers, I recommend checking out the [PWA feature plugin](https://wordpress.org/plugins/pwa/) as it includes some of these things I mentioned already, just for other parts of WordPress. For example, it has background sync for comments and support for a dedicated offline template that is shown when there's no cached content.

aduth commented:
Thanks @swissspidy . These all sound like worthwhile explorations. In the short term, I think we can still pursue the approach in this pull request independently, and keep #6322 open for further discussion and enhancements.

aduth committed:
Editor: Fix comment describing intent of localStorage test

aduth committed:
Editor: Implement localStorage autosave restore

aduth commented:
I've finished fleshing out the remaining functionality, and have updated the original comment with an animation, implementation notes, and testing instructions.

aduth committed:
Editor: Construct autosave edits directly from parse

JSON.stringify would not include properties with undefined values, so it is not necessary to omit them explicitly

mtias commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -76,6 +77,7 @@ function Layout( {
 			<BrowserURL />
 			<UnsavedChangesWarning />
 			<AutosaveMonitor />
+			<LocalAutosaveMonitor />
```
Would it make sense to combine some of these monitors?

ellatrix replied:
I wonder as well. Would it make sense to only save locally if saving otherwise fails?

aduth replied:
> Would it make sense to combine some of these monitors?

As implemented, the `LocalAutosaveMonitor` is effectively an extension of the base `AutosaveMonitor` behavior.

aduth replied:
> I wonder as well. Would it make sense to only save locally if saving otherwise fails?

We could. The local autosave occurs more frequently than the server-persisted autosave (by default at least), so I'd assume there's technically a potential for data loss in that ~45s gap. I dunno that it's worth the extra complexity to determine whether a failure occurred.

mcsf replied:
Yes, let's save locally regardless of remote saving. Plus, now that we are to switch to `sessionStorage`, we have no real reason to worry about wasting storage.

mtias commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+ *
+ * @return {Function} Callback function.
+ */
+function useAutosaveSaveCallback() {
```
`useAutosaveCallback`?

aduth replied:
> Should this be `useAutosaveCallback`?

Yeah, it might read a little better. I think I chose it on the basis of being a specific sort of save callback.

>If we are explicitly including the local storage action, maybe `useAutosaveStore` could work too.

Sure. We might want to consider if we want any naming conventions around values assigned from hooks, e.g. in this case `useCallback`.

mcsf replied:
> Sure. We might want to consider if we want any naming conventions around values assigned from hooks, e.g. in this case `useCallback`.

That's good enough for me, at least until we have proper conventions. Let's stick with `useAutosaveCallback`.

talldan commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+
+	return (
+		<AutosaveMonitor
+			interval={ AUTOSAVE_INTERVAL_SECONDS }
```
This seems to require that the user is idle for 15 seconds before it'll trigger a save.

Since this is a last resort, I wonder if there's also an option to use additional strategies for saving local autosaves, like the beforeUnload handler.

westonruter replied:
Given that `localStorage` is a synchronous API, should [`requestIdleCallback`](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback) be used if available.

ellatrix replied:
Perhaps related: https://github.com/WordPress/gutenberg/pull/16490/files#r303243055

Maybe we should only save locally if you cannot save otherwise? Would it make more sense then to implement it in a different layer of the application?

aduth replied:
> Since this is a last resort, I wonder if there's also an option to use additional strategies for saving local autosaves, like the beforeUnload handler.

I'd agree on this, and the fact that we should probably just schedule (or at least check for changes to justify) the autosaves to occur _every_ 15 seconds, not just 15 seconds after the last change.

> Given that `localStorage` is a synchronous API, should [`requestIdleCallback`](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback) be used if available?

It could be worth exploring. I'd question whether it would make a measurable difference, however (related: #16657).

talldan reviewed (commented):
Seems to be on the right track. 👍

talldan commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+
+		if ( ! hasDifference ) {
+			// If there is no difference, it can be safely ejected from storage.
+			window.localStorage.removeItem( 'post_' + postId );
```
This would mean that if a second user logs in to the same browser, they can restore an autosave made by the first user.

Not sure if that's an issue, but I know the current autosave system stores revisions by different users separately. One way to align with that but still keep the same localstorage key is to save the userId in the payload.

aduth replied:
> This key format would mean that if a second user logs in to the same browser, they can restore an autosave made by the first user.

Based on other comments, I think `sessionStorage` may be enough to resolve this.

Though technically it could still be an issue if the user logs out of the current user and into another user while still within the same tab / session. I'd be curious to know if / how this was addressed in the classic editor.

adamsilverstein replied:
Currently in core we clear autosave data from session storage when the user logs out: https://github.com/WordPress/wordpress-develop/blob/6dad32d2aed47e6c0cf2aee8410645f6d7aba6bd/src/wp-login.php#L103

talldan commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+
+		const noticeId = uniqueId( 'wpEditorAutosaveRestore' );
+
+		createWarningNotice( __( 'The backup of this post in your browser is different from the version below.' ), {
```
Just thinking of all the weird and wonderful edge cases that might happen. I think there is a chance that a user might create a revision for a post on a second computer, and then when returning to the first computer they'd be presented with both this notice and the autosave notice.

One option would be to timestamp local saves, and then only present the user with the notice for the most recent save whether it's a local or a server stored save.

mcsf commented:
One thing that this PR doesn't need to address but which will need a solution to in the future is cleaning up after deleted posts.

In the following screencast, a draft with changes sitting in localStorage is deleted, leaving behind its entry in localStorage.

![localstorage-delete-cleanup](https://user-images.githubusercontent.com/150562/61451117-f8544800-a958-11e9-81eb-f6cfccefc551.gif)

westonruter commented:
This probably won't matter so much, but I wanted to mention: `localStorage` is a synchronous API. If many writes are being done with large bodies of content into `localStorage`, it is possible that this could degrade performance in the editor. See https://developers.google.com/web/updates/2019/03/kv-storage

Also, `localStorage` would persist in the browser even when the user is logged-out. On a shared computer, this would allow leakage of data, so ideally the data would be encrypted with some user-specific secret. Maybe this is overkill.

See also `kv-storage` for async storage: https://github.com/WICG/kv-storage

ellatrix commented:
WordPress Core currently uses `sessionStorage` to back up the post content because of the security issues with `localStorage`.

ellatrix commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+	} ) );
+
+	return useCallback( () => {
+		window.localStorage.setItem( 'post_' + postId, JSON.stringify( {
```
We also should use `sessionStorage`. :)

mtias replied:
Yes, let's switch to `sessionStorage`-

aduth replied:
> The REST autosave system has some checks in place to make sure either post_title or content is defined before saving. Worth replicating that here?

Could you point me to what it is you're referring to on this?

talldan replied:
Yep, it's `isEditedPostSaveable`:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/selectors.js#L458-L462

Which is used to return early from `savePost` in the event there aren't changes to any of those autosave fields:
https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/actions.js#L439

adamsilverstein replied:
I expect the overhead for saving to sessionStorage is very low, possibly lower than retrieving and comparing to the old data?

mcsf replied:
> I expect the overhead for saving to sessionStorage is very low, possibly lower than retrieving and comparing to the old data?

Agreed, unless we measure otherwise.

aduth commented on packages/editor/src/components/autosave-monitor/index.js:
```diff
@@ -58,25 +58,29 @@ export class AutosaveMonitor extends Component {
-		autosave: dispatch( 'core/editor' ).autosave,
+	withDispatch( ( dispatch, ownProps ) => ( {
+		autosave() {
+			const { autosave = dispatch( 'core/editor' ).autosave } = ownProps;
```
> An option might be to extract a more generic `BaseAutosaveMonitor` component.

Hmm, I see where you're coming from. Some of this is inherited from the previous behavior that the editor has some understanding of what an "autosave" action means (the server-persisted autosave).

Funny though that your recommendation is essentially the opposite of @mtias at https://github.com/WordPress/gutenberg/pull/16490#discussion_r303243055 😄

talldan replied:
😄 Hah, true.

Combining them would allow more coordination between the two autosave systems, which might be beneficial.

aduth commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+			actions: [
+				{
+					label: __( 'Restore the backup' ),
+					onClick() {
```
>The hard part in terms of the user experience here is that this action might destroy more of the user's content if the user has added more to the post since the last local autosave. Although they can use undo, I don't think that's always apparent.
>
>The other autosave system displays a diff before giving the option to apply changes. Although it's a bit of extra work, it could be worth trying. I believe there's a diff library already in use for block validation.

By other autosave, I assume you mean the revisions? There was also a localStorage/sessionStorage-based solution in the classic editor which, as far as I recall, took no such precautions.

I'm leaning more on the end that Undo should serve as the intended option should the changes not be what the user was expecting, maybe at least as a first pass.

talldan replied:
Yep, I mean revisions. Wouldn't want to add too much more to this PR, so if considering that part separately is an option, that sounds good.

adamsilverstein replied:
Good point, in the classic editor we show the user the following message when an autosave is present in sessionStorage:

![image](https://user-images.githubusercontent.com/2676022/63546150-52ab7000-c4e6-11e9-9748-e87c5593a728.png)

Can we match something like that messaging here? Is this still true in Gutenberg, can the user use 'undo' to undo the restore?

> The other autosave system displays a diff before giving the option to apply changes. Although it's a bit of extra work, it could be worth trying. I believe there's a diff library already in use for block validation.

Great idea! I suggest we revisit this when we develop a more full fledged revision UI into Gutenberg.

aduth commented:
Thanks all for the feedback!

A few specific action items I've observed:

- Always run the autosave procedure _every_ 15 seconds, not 15 seconds after the last change
   - Still try to save to storage only if a change has actually occurred
   - Explore whether `requestIdleCallback` is a reasonable improvement to avoid blocking operation
- Add some more through inline code documention (e.g. `lastDidSave`)
- Convert to use `sessionStorage` as a way to avoid accumulating storage and potential for lingering storage from other users
   - Should confirm that this data is in-fact purged upon logging out, or that an equivalent measure is taken to avoid access by other users even within the same storage

paaljoachim commented:
On occasion I have some hickups - as in a slow internet causing this: https://github.com/WordPress/gutenberg/issues/17015

It would be very helpful to have this PR merged! Thanks.

mcsf committed:
Rename to useAutosaveCallback

mcsf committed:
Adopt wp-autosave- schema for sessionStorage keys

Keys are crucially prefixed with 'wp-autosave-' so that wp-login.php's
`loggedout` handler can clear sessionStorage of any user-private
content. See:

https://github.com/WordPress/wordpress-develop/blob/6dad32d2aed47e6c0cf2aee8410645f6d7aba6bd/src/wp-login.php#L103

mcsf committed:
LocalAutosaveMonitor: Switch from local to sessionStorage

mcsf commented:
As Andrew is enjoying some time off, I've revived this branch and applied some of the feedback that's been provided. A couple of notes:

* In terms of terminology, I've kept `LocalAutosave` even though "local storage" became "session storage" throughout the diff. I think LocalAutosave is still more descriptive and perfectly apt, as it doesn't imply usage of `localStorage`, but rather only locality, which `sessionStorage` also provides.
* @adamsilverstein's tip https://github.com/WordPress/gutenberg/pull/16490#discussion_r316861235 was quite useful. I did however find a strange situation where attempting to log out while there are unsaved edits in the editor leads to deleting the autosaves from `sessionStorage` even if you cancel navigation in the `onbeforeunload` confirmation:

![sessionStorage-logout](https://user-images.githubusercontent.com/150562/64047891-f7bdec80-cb67-11e9-8e72-a17472a8885e.gif)

I don't think this is something to address in this PR, but it's something to be aware of.

mcsf commented:
Updated PR title and description, wherein I've added a checklist.

mcsf committed:
LocalAutosaveMonitor: Wrap saving in requestIdleCallback

mcsf commented:
Since rebasing against `master`, I notice that `AutosaveMonitor`'s `componentDidUpdate` is no longer firing "often enough", i.e. as the user edits their post. As a result, the first autosave happens as expected, but no other autosaves happen — until the user explicitly saves the draft, thereby touching `AutosaveMonitor`'s props `isAutosaveable` and `isDirty`.

Presumably, this is a consequence of `getReferenceByDistinctEdits` no longer working as intended. @epiqueras, think this is related to some of your recent work? What do you suggest?

mcsf commented:
Btw, context for my last comment and `getReferenceByDistinctEdits`: #10103

epiqueras commented:
```js
export const getReferenceByDistinctEdits = createSelector(
	() => [],
	( state ) => [ state.editor ],
);
```

Yes, `state.editor` does not exist anymore.

What do you think about changing that to infer the same thing from the length/size of the undo stack?

mcsf commented:
> infer the same thing from the length/size of the undo stack?

Why not infer from the actual reference of (core's) `state.undo`, similarly to what is currently done with (editor's) `state.editor`?

epiqueras commented:
>Why not infer from the actual reference of (core's) state.undo, similarly to what is currently done with (editor's) state.editor?

That would also change when an edit that does not create a new undo level is made, a "flattened edit".

We probably don't want that behavior since those edits are not persisted and this is for the autosave monitor.

mcsf committed:
getReferenceByDistinctEdits: move to core-data

* Base implementation on undo state of core entities
* Expand unit tests

mcsf commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,169 @@
+		didSave: select( 'core/editor' ).didPostSaveRequestSucceed(),
+	} ) );
+
+	const lastDidSave = useRef( didSave );
```
Indeed. Now that this PR had been rebased on top of the merged core entities changes (#17368), `useAutosavePurge` no longer works, as `didSave` is defaulting to true.

Neither scenario is wrong, and both are unspecified behaviour. :)

Instead, this hook will need to observe some additional state, perhaps publish/draft status, or just dirtiness. @epiqueras, considering how entities now work, does any of these sound better to explore than the other?

epiqueras replied:
I think this is what you want?

```js
const { isSavingEntityRecord, getLastEntitySaveError } = select('core/data')
const didSave =
	isSavingEntityRecord(...args) === false && !getLastEntitySaveError(...args)
```

Note that `isSavingEntityRecord(...args)` is undefined before the first save, hence the `=== false`.

mcsf replied:
I pushed something that seems to be working that's based on `isDirty` instead. If it won't cover our needs we could look at this.

mcsf committed:
Fix useAutosaveNotice by observing isEditedPostDirty

mcsf committed:
AutosaveMonitor: Support throttle apart from debounce

mcsf committed:
Add e2e tests for sessionStorage-based autosave

gziolo commented on packages/core-data/src/test/selectors.js:
```diff
@@ -275,3 +276,45 @@ describe( 'getCurrentUser', () => {
+	describe( 'when adding edits', () => {
+		it( 'should return referentially different values across changing states', () => {
+			const beforeState = { undo: [ {} ] };
+			beforeState.undo.offset = 0;
```
Can you elaborate what's happening here? I can't figure out myself which I see as a good reason to fix with an inline comment.

mcsf replied:
The `undo` reducer manually manages this `offset` property:

https://github.com/WordPress/gutenberg/blob/52ca63ee461848a742addac3e78a9fc0f637b5b5/packages/core-data/src/reducer.js#L304-L313

which is then used in selectors:

https://github.com/WordPress/gutenberg/blob/52ca63ee461848a742addac3e78a9fc0f637b5b5/packages/core-data/src/selectors.js#L293-L307

This particular case, as is in its name, mimics the addition of an edit state. So, aside from expanding the `undo` array, we also bump `offset` to point to the latest state.

The second test mimics the undo by just lowering the offset. Although I now see that there should be two objects in `undo` instead of one, for fidelity. :)

gziolo replied:
It sounds like a very fragile construct but I guess this is what it is. I hope all those tests where undo state is manually tweaked won't produce too many issues when we try to change its internal shape in the future. 

I assume it's good as is :)

gziolo commented on packages/editor/src/components/autosave-monitor/index.js:
```diff
@@ -42,12 +42,25 @@ export class AutosaveMonitor extends Component {
+		// The `shouldThrottle` props allows overriding this behaviour, thus
+		// making the autosave action "throttled".
+		if ( ! shouldThrottle ) {
+			clearTimeout( this.pendingSave );
```
Should `clearTimeout` call be guarded here to avoid executing this block when `this.pendingSave` is not defined?

gziolo commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,191 @@
+	}, [ isDirty ] );
+}
+
+function LocalAutosaveMonitor() {
```
It would be great to add some docs as a follow-up PR since it's a public API :)

gziolo commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,191 @@
+		<AutosaveMonitor
+			interval={ AUTOSAVE_INTERVAL_SECONDS }
+			autosave={ autosave }
+			shouldThrottle
```
Why do we pass `shouldThrottle` always set to `true`? Well, the question is if we really need all the logic which covers the case where it would be `false`.

mcsf replied:
The point was to leave server-side autosave untouched. Presumably, since there's a network transaction and server-side execution, we'd want to be more conservative with autosave — hence defaulting to its debounced action. With local autosave we can be more eager. What do you think?

gziolo replied:
I didn't notice we are reusing the existing component. It's all good. I wasn't sure why we even `shouldThrottle` have to pass this value in the first place. Now, it's all clear.

gziolo commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,191 @@
+ * Keys are crucially prefixed with 'wp-autosave-' so that wp-login.php's
+ * `loggedout` handler can clear sessionStorage of any user-private content.
+ *
+ * @see https://github.com/WordPress/wordpress-develop/blob/6dad32d2aed47e6c0cf2aee8410645f6d7aba6bd/src/wp-login.php#L103
```
Nice, this is helpful 👍

gziolo commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,191 @@
+	return useCallback( () => {
+		const saveToSessionStorage = () => {
+			window.sessionStorage.setItem( postKey( postId ), JSON.stringify( {
+				post_title: getEditedPostAttribute( 'title' ),
```
What about other stuff like:
- taxonomy
- categories
- featured image

and so on?

ellatrix replied:
Worth noting that this wasn't saved in the classic editor, but that doesn't mean we should do the same. Just noting.

mcsf replied:
Yeah. Grzegorz and I also talked a little about this:

* There's also a management of expectations to play with: the more data we autosave, the greater the expectations that this is a fully fledged offline support feature, which it isn't.

* I think backing up post content really is 90% of the feature, and for an introductory PR this is more than fine. The fact that Classic did the same validates this.

* As Gutenberg evolves to support more than just posts with content and into the realm of full-site editing, a lot of this might become a lot more ambitious if we're dealing with multiple entities to save in a single editor. (When we get there we can try autosaving around each `BlockEditor` and see where it takes us.)

For now, let's keep it focused on these three bits.

gziolo commented on packages/editor/src/components/local-autosave-monitor/index.js:
```diff
@@ -0,0 +1,191 @@
+			} ) );
+		};
+		requestIdleCallback( saveToSessionStorage );
+	}, [ postId, getEditedPostAttribute ] );
```
`getEditedPostAttribute` - we put this action as a dependency here, but we don't do it for all actions in `useAutosaveNotice`. The question is why it is necessary here or why do we miss all those actions for `useEffect` inside `useAutosaveNotice`.

Is it because of `useAutosaveNotice` should be triggered only once per post id?

mcsf replied:
Per individual chat, let's drop `getEditedPostAttribute` from the deps. We don't yet have strong conventions on how to deal with this in React hooks.

If we were very strict about this we would list every single reference used in the closure, but the references in our case are all static constant functions, either directly (defined in the module or imported) or indirectly (selectors obtained via `select`). I don't feel strongly, but I'd err toward keeping references out for now and adding them if we tighten our expectations here.

gziolo commented on packages/e2e-tests/specs/autosave.test.js:
```diff
@@ -0,0 +1,98 @@
+	pressKeyWithModifier,
+} from '@wordpress/e2e-test-utils';
+
+const AUTOSAVE_INTERVAL_SECONDS = 15;
```
That's a big number of seconds to wait in the test, to be fair. Can we set something lower through some sort of override? In other words, can we avoid using sleep as much as it is possible?

mcsf replied:
We could bring it down to 5, if that makes sense as a user experience.

It might also be beneficial to make this an editor setting so it could be customised. Would this be a requirement for this particular PR?

ellatrix replied:
How slow would setting session storage be for large posts?

gziolo commented on packages/e2e-tests/specs/autosave.test.js:
```diff
@@ -0,0 +1,98 @@
+		await createNewPost();
+	} );
+
+	it( 'should save to sessionStorage', async () => {
```
Those tests look great. We need to optimize them time-wise only,

mcsf commented:
It needs to be profiled, and we also need to benchmark with the perf e2e suite.

gziolo reviewed (approved):
This works great, let's move forward with it once we add back deprecated message for the removed selector with the proper warning printed to the console. See: https://github.com/WordPress/gutenberg/pull/16490#discussion_r324490171

As a follow-up tasks we should do the following:
- Add documentation for the newly introduced component
- Make the local autosave interval configurable so sites could tweak it and at the same time to speed up e2e tests
- Let's add more e2e tests which cover other use cases like (mostly to prevent regressions):
  - logging out - which wipes the saved post
  - ensuring that the saved content doesn't get loaded in the other post

mcsf committed:
Address feedback

mcsf committed:
Make localAutosaveInterval a setting to speed up e2e tests

mcsf commented:
A performance benchmark:

## master

	Average time to load: 10858ms
	Average time to DOM content load: 10023ms
	Average time to type character: 70.515ms
	Slowest time to type character: 132ms
	Fastest time to type character: 56ms

	Average time to load: 10793ms
	Average time to DOM content load: 9916ms
	Average time to type character: 67.465ms
	Slowest time to type character: 142ms
	Fastest time to type character: 59ms

## local autosave, no editor setting

	Average time to load: 11169ms
	Average time to DOM content load: 10211ms
	Average time to type character: 71.01ms
	Slowest time to type character: 190ms
	Fastest time to type character: 61ms
			
	Average time to load: 10685ms
	Average time to DOM content load: 9811ms
	Average time to type character: 72.015ms
	Slowest time to type character: 156ms
	Fastest time to type character: 58ms

## local autosave, with editor setting

	Average time to load: 9941ms
	Average time to DOM content load: 9647ms
	Average time to type character: 72.995ms
	Slowest time to type character: 157ms
	Fastest time to type character: 62ms

	Average time to load: 10343ms
	Average time to DOM content load: 10041ms
	Average time to type character: 67.45ms
	Slowest time to type character: 160ms
	Fastest time to type character: 53ms

mcsf commented on packages/editor/src/store/selectors.js:
```diff
@@ -230,29 +230,6 @@ export const getPostEdits = createRegistrySelector( ( select ) => ( state ) => {
- *
- * @return {*} A value whose reference will change only when an edit occurs.
- */
-export const getReferenceByDistinctEdits = createSelector(
```
#17458

aduth commented on packages/e2e-tests/specs/autosave.test.js:
```diff
@@ -0,0 +1,107 @@
+	beforeEach( async () => {
+		await clearSessionStorage();
+		await createNewPost();
+		await setLocalAutosaveInterval( AUTOSAVE_INTERVAL_SECONDS );
```
Should there be an associated tear down to this, to avoid the state bleeding over into other tests run in the same container?

Not sure it would have any noticeable impact since ideally those tests would be resilient regardless of an autosave interval, but seems like an undesirable side-effect in any case.

aduth replied:
I guess it might depend if `setLocalAutosaveInterval` is something which would persist outside the current page. If most tests would create a new post and thus navigate away, it might not be a problem.

Then again, it's long been a desire of mine that `createNewPost` shouldn't have to trigger a full page reload (see #18668).
